beefy-gadget/src/dkg.rs

```diff
@@ -0,0 +1,184 @@
+use bincode;
+use codec::{Decode, Encode};
+use log::{debug, error, info, trace, warn};
+use round_based::{IsCritical, Msg, StateMachine};
+use sp_keystore::{Error, SyncCryptoStore};
+use sp_runtime::traits::{Block, Hash, Header, NumberFor};
+use std::sync::Arc;
+
+pub use multi_party_ecdsa::protocols::multi_party_ecdsa::gg_2020::{party_i::*, state_machine::keygen::*};
+
+use crate::error::MPCError;
+use beefy_primitives::crypto::Public;
+
+#[derive(Debug, Decode, Encode)]
+#[cfg_attr(feature = "scale-info", derive(scale_info::TypeInfo))]
+pub enum DKGType {
+	MultiPartyECDSA,
+}
+
+/// Gossip engine webb messages topic
+pub(crate) fn webb_topic<B: Block>() -> B::Hash
+where
+	B: Block,
+{
+	<<B::Header as Header>::Hashing as Hash>::hash(b"webb")
+}
+
+/// WEBB DKG (distributed key generation) message.
+///
+/// A vote message is a direct vote created by a WEBB node on every voting round
+/// and is gossiped to its peers.
+#[derive(Debug, Decode, Encode)]
+#[cfg_attr(feature = "scale-info", derive(scale_info::TypeInfo))]
+pub struct DKGMessage<Public> {
+	/// Node authority id
+	pub id: Public,
+	/// DKG protocol type identifier
+	pub dkg_type: DKGType,
+	/// DKG message contents
+	pub message: Vec<u8>,
+}
+
+/// Multi party ECDSA trait for the keystore.
+pub trait MultiPartyECDSAKeyStore: SyncCryptoStore {
+	/// Generate keys for a new participant.
+	fn generate(&self, index: usize) -> Result<Keys, Error>;
+	/// Refresh the keys for a participant.
+	fn refresh(&self, index: usize) -> Result<Keys, Error> {
+		self.generate(index)
+	}
+}
+
+/// A pointer to a keystore.
+pub type MultiPartyCryptoStorePtr = Arc<dyn MultiPartyECDSAKeyStore>;
+
+pub struct MultiPartyECDSASettings {
+	pub threshold: usize,
+	pub parties: usize,
+	pub accepted: bool,
+	pub party_index: usize,
+	pub keygen: Keygen,
+	pub local_key: Option<LocalKey>,
+}
+
+impl MultiPartyECDSASettings {
+	pub fn new(threshold: usize, parties: usize, party_index: usize) -> Result<Self, MPCError> {
+		// TODO u16 limit(=65k) is good enough, change type to u16?
```

Comment:
Yea, we should change this.

Reply:
Changed it.

Reply:
RUST_LOG=webb=trace ./target/debug/beefy-node --tmp --alice --port 30334 --ws-port 9944 &> alice.log